hypershift-operator/controllers/nodepool/kubevirt/imagecacher.go

```diff
@@ -0,0 +1,263 @@
+	bootImagePVCLabelRoleNameForPatch = "hypershift.openshift.io~1kubevirt-boot-image-role" // Can't use slashes as keys in jsonpatch. Replace them with "~1"
+	bootImagePVCLabelRoleValue        = "kv-boot-image-cache"
+	bootImageNamePrefix               = bootImagePVCLabelRoleValue + "-"
+
+	maxK8sNameLen = 63
+
+	// A CDI annotation for DataVolume, to not wait to first customer, but start importing immediately.
+	// originally defined in CDI.
+	cdiImmediateBindingAnnotation = "cdi.kubevirt.io/storage.bind.immediate.requested"
+)
+
+// BootImage represents the KubeVirt boot image. It responsible to hold cache the boot image and to generate its
+// reference to be used by the node templates.
+type BootImage interface {
+	// CacheImage creates a PVC to cache the node image.
+	CacheImage(ctx context.Context, cl client.Client, nodePool *hyperv1.NodePool) error
+	getDVSourceForVMTemplate() *v1beta1.DataVolumeSource
+}
+
+// containerImage is the implementation of the BootImage interface for container images
+type containerImage struct {
+	name string
+}
+
+func newContainerBootImage(imageName string) *containerImage {
+	return &containerImage{
+		name: imageName,
+	}
+}
+
+func (containerImage) CacheImage(_ context.Context, _ client.Client, _ *hyperv1.NodePool) error {
+	return nil // no implementation
+}
+
+func (ci containerImage) getDVSourceForVMTemplate() *v1beta1.DataVolumeSource {
+	pullMethod := v1beta1.RegistryPullNode
+	return &v1beta1.DataVolumeSource{
+		Registry: &v1beta1.DataVolumeSourceRegistry{
+			URL:        &ci.name,
+			PullMethod: &pullMethod,
+		},
+	}
+}
+
+// qcowImage is the implementation of the BootImage interface for QCOW images
+type qcowImage struct {
+	name      string
+	hash      string
+	namespace string
+	dvName    string
+	log       logr.Logger
+}
+
+func newQCOWBootImage(name, hash, namespace string, logger logr.Logger) *qcowImage {
+	return &qcowImage{
+		name:      name,
+		hash:      hash,
+		namespace: namespace,
+		log:       logger,
+		dvName:    getDvName(hash),
+	}
+}
+
+func (qi qcowImage) CacheImage(ctx context.Context, cl client.Client, nodePool *hyperv1.NodePool) error {
+	kvPlatform := nodePool.Spec.Platform.Kubevirt
+
+	objectKey := client.ObjectKey{Namespace: qi.namespace, Name: qi.dvName}
+
+	dvFound, err := dvExists(ctx, cl, objectKey)
+	if err != nil {
+		return fmt.Errorf("failed to read the DataVolume; namespace: %s; name: %s; %w", qi.namespace, qi.dvName, err)
+	}
+
+	pvc := &corev1.PersistentVolumeClaim{}
+	err = cl.Get(ctx, objectKey, pvc)
+	if errors.IsNotFound(err) { // the PVC wasn't found
+		if !dvFound {
+			qi.log.Info("couldn't find boot image cache DataVolume; creating it...", "DataVolume Name", qi.dvName)
+			return qi.createDVForCache(ctx, cl, kvPlatform)
+		} else {
+			qi.log.Info("found the boot image cache DataVolume. The PVC should be created shortly")
+		}
+		return nil // the DV was already created, so we just need to wait for the PVC to be created.
+	}
```

Comment:
It's possible for us to only work with DataVolumes entirely here, and I think that will simplify our lives a bit. These cached images are not user facing.

add `"cdi.kubevirt.io/storage.deleteAfterCompletion": "false",` to the datavolume's annotations and the datavolume will stay around. That way we don't have to do the dance between the datavolume and the pvc... also we can just use the annotations/labels directly  on the DV and not worry about when (or if) they propagate  correctly to the PVC.